### ISSUE TITLE
pptx get really slow when generate many graphs

### ISSUE THREAD

**Issue by slfan2013** [davidgohel/officer#95]
**ISSUE:**
the pptx get really slow when generating many plots. First ~10 plots is fast and the more slides added, the slower. Plotting 100 graphs using `officer` costs more than 60 secs. Using another package, `ReporteRs`, only takes 2 secs.

Please run following code,
```
library(officer,dplyr, rvg)
start = Sys.time()
pptx = read_pptx()
for(i in 1:100){
  print(i)
  pptx = pptx%>% add_slide(layout = "Title and Content", master = "Office Theme") %>%
    ph_with_vg(code = {
      boxplot(Sepal.Length~Species, data = iris)
    }, type = "body", width = 8, height = 6, offx = 0, offy = 0)
}
pptx %>% print(target = "100 boxplot using officer.pptx") %>% invisible()
end = Sys.time()
end-start #the time difference is 1 min
```
The above code took more than 60 sec to finish 100 plots. However, the package `ReporteRs` only take 2 secs. Please see following code of generating same pptx using `ReporteRs`,

```
library(ReporteRs)
start = Sys.time()
doc = pptx( )
for(i in 1:100){
  print(i)
  doc = addSlide(doc, slide.layout = "Title and Content")
  doc = addPlot(doc, fun = function() {
    boxplot(Sepal.Length~Species, data = iris)
  },
  vector.graphic = TRUE, width = 8, height = 6)
}

writeDoc(doc, file = "100 boxplot using ReporteRs.pptx")
end = Sys.time()
end-start #Time difference of 1.932549 secs
```

I don't want to switch to `ReporteRs` because it depends on Java. I appreciate if you could fix this.

**Comment by davidgohel** [comment]
Thanks

There will be work on optimizing the production of pptx but not now.

**Comment by davidgohel** [comment]
@md0u80c9 @slfan2013 

I refactored the pptx code and the result is good. On my computer (mac book pro 16 Go):

* 100 slides with officer code: `7.625977 secs`
* 100 slides with ReporteRs code: `3.663771 secs`
* 10000 slides with officer code: `24.03007 mins`
* 10000 slides with ReporteRs code: interupted at `4.770214` because of `java.lang.OutOfMemoryError: GC overhead limit exceeded`



2 unnecessary function calls caused many performance issue. Also the current slide was saved each time its content was modified, they are now saved only once when writing the `pptx` output file.

**Comment by md0u80c9** [comment]
This is awesome news! I'll give this a test later today (once I've resolved a particularly irritating bug in my own code!)
